api/utils/sshutils/conn_test.go

```diff
@@ -0,0 +1,173 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sshutils
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/pem"
+	"net"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/gravitational/teleport/api/constants"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/crypto/ssh"
+)
+
+type server struct {
+	listener net.Listener
+	config   *ssh.ServerConfig
+	handler  func(*ssh.ServerConn)
+	t        *testing.T
+	mu       sync.RWMutex
+	closed   bool
+
+	cSigner ssh.Signer
+	hSigner ssh.Signer
+}
+
+func (s *server) Run() {
+	for {
+		conn, err := s.listener.Accept()
+
+		s.mu.RLock()
+		if s.closed {
+			s.mu.RUnlock()
+			return
+		}
+		s.mu.RUnlock()
+
+		assert.NoError(s.t, err)
+
+		go func() {
+			defer conn.Close()
+			sconn, _, _, err := ssh.NewServerConn(conn, s.config)
+			assert.NoError(s.t, err)
```

Comment:
`assert` or `require` are not going to work. You can only make assertion from the main goroutine. I think that the best that you can do there is to create an `chan error` and propagate the error from the goroutine. Example https://github.com/gravitational/teleport/blob/663e3d04c545de0dc8275f173ef9010ecc6affbe/lib/srv/db/access_test.go#L1196-L1215

Reply:
Ah didn't realize this thanks. Updated here [a59074c](https://github.com/gravitational/teleport/pull/11875/commits/a59074c1320f721943e2acd3ddc04117825547de)

Reply:
I changed this to use a channel and now I understand not to use `require` since it calls FailNow under the hood. But using `assert` calls `Fail` under the hood which waits for the test to return so it sounds like that should be safe to call from a goroutine as long as we wait for goroutines to finish before the test ends.

I don't plan on changing this back to using assert just trying to understand if it would work.